src/enzyme_ad/jax/Passes/AffineCFG.cpp

```diff
@@ -4325,6 +4326,254 @@ struct SinkStoreInIf : public OpRewritePattern<scf::IfOp> {
+      return rewriter.notifyMatchFailure(
+          loadOp, "not dependent on a conditional result");
+
+    auto toLift = llvm::filter_to_vector(backwardSlice, [&](Operation *op) {
+      return dominance.properlyDominates(conditional, op);
+    });
+    if (llvm::any_of(toLift, [](Operation *op) { return !isPure(op); }))
+      return rewriter.notifyMatchFailure(loadOp,
+                                         "non-pure operation on the path");
+
+    auto cloneIntoBlock = [&](unsigned blockNum) {
+      IRMapping mapping;
+      Block &targetBlock = conditional->getRegion(blockNum).front();
+      mapping.map(conditional->getResults(),
+                  targetBlock.getTerminator()->getOperands());
+      rewriter.setInsertionPoint(targetBlock.getTerminator());
+      for (Operation *op : toLift) {
+        rewriter.clone(*op, mapping);
+      }
+      Operation *clonedLoad = rewriter.clone(*loadOp.getOperation(), mapping);
+      return clonedLoad;
+    };
+
+    Operation *thenLoad = cloneIntoBlock(0);
+    Operation *elseLoad = cloneIntoBlock(1);
+
+    SmallVector<Type> types = llvm::to_vector(conditional->getResultTypes());
+    llvm::append_range(types, thenLoad->getResultTypes());
+    SmallVector<std::unique_ptr<Region>> regions;
+    regions.emplace_back(new Region);
+    regions.emplace_back(new Region);
+    rewriter.setInsertionPoint(conditional);
+    Operation *newConditional = rewriter.create(
+        conditional->getLoc(), conditional->getName().getIdentifier(),
+        conditional->getOperands(), types, conditional->getAttrs(),
+        BlockRange(), regions);
+
+    auto inlineBody = [&](unsigned regionNum, Operation *loadOp) {
+      rewriter.inlineRegionBefore(conditional->getRegion(regionNum),
+                                  newConditional->getRegion(regionNum),
+                                  newConditional->getRegion(regionNum).begin());
+
+      Operation *terminator =
+          newConditional->getRegion(regionNum).front().getTerminator();
+      SmallVector<Value> operands = llvm::to_vector(terminator->getOperands());
+      llvm::append_range(operands, loadOp->getResults());
+      rewriter.setInsertionPoint(terminator);
+      Operation *newTerminator = rewriter.create(
+          terminator->getLoc(), terminator->getName().getIdentifier(), operands,
+          terminator->getResultTypes(), terminator->getAttrs(),
+          terminator->getSuccessors());
+      rewriter.replaceOp(terminator, newTerminator);
+    };
+
+    inlineBody(0, thenLoad);
+    inlineBody(1, elseLoad);
+
+    unsigned numLoadResults = loadOp->getNumResults();
+    rewriter.replaceOp(loadOp,
+                       newConditional->getResults().take_back(numLoadResults));
+    rewriter.replaceOp(conditional,
+                       newConditional->getResults().drop_back(numLoadResults));
+    return success();
+  }
+};
+
+template <typename Derived, typename BinOp>
+struct FoldAffineApplyBase : public OpRewritePattern<BinOp> {
```

Comment:
this should be handled by https://github.com/EnzymeAD/Enzyme-JAX/blob/e4438171e9a5bc5c62e906f2ee0be3b3e035b145/src/enzyme_ad/jax/Passes/AffineCFG.cpp#L1003 no?

Reply:
This goes into the opposite direction: we now fold _users_ of affine apply.

Reply:
ah got it